BJS2-77994 Карьера

dextomi opened the pull request:
Реализация основных операций по карьере

Ramil23-bit committed:
added data about work schedule, first part

phareebo committed:
задала базовый url-путь с версионированием для всех контроллеров

phareebo committed:
задала базовый url-путь с версионированием для всех контроллеров

phareebo committed:
задала базовый url-путь с версионированием для всех контроллеров

phareebo committed:
задала базовый url-путь с версионированием для всех контроллеров

phareebo committed:
задала базовый url-путь с версионированием для всех контроллеров

phareebo committed:
задала базовый url-путь с версионированием для всех контроллеров

phareebo committed:
метод создания события, exception & error handlers, fix cyclic dependence

phareebo committed:
метод создания события, exception & error handlers, fix cyclic dependence

phareebo committed:
метод создания события, exception & error handlers, fix cyclic dependence

phareebo committed:
метод создания события, exception & error handlers, fix cyclic dependence

phareebo committed:
метод создания события, exception & error handlers, fix cyclic dependence

phareebo committed:
метод создания события, exception & error handlers, fix cyclic dependence

phareebo committed:
Revert "set base path for all controllers"

This reverts commit d03e9aedbb9a509783e41101cbd8649f4e851341.

phareebo committed:
Revert "set base path for all controllers"

This reverts commit d03e9aedbb9a509783e41101cbd8649f4e851341.

phareebo committed:
Revert "set base path for all controllers"

This reverts commit d03e9aedbb9a509783e41101cbd8649f4e851341.

phareebo committed:
Revert "set base path for all controllers"

This reverts commit d03e9aedbb9a509783e41101cbd8649f4e851341.

phareebo committed:
Revert "set base path for all controllers"

This reverts commit d03e9aedbb9a509783e41101cbd8649f4e851341.

phareebo committed:
Revert "set base path for all controllers"

This reverts commit d03e9aedbb9a509783e41101cbd8649f4e851341.

phareebo committed:
выпилила exception handler

phareebo committed:
выпилила exception handler

phareebo committed:
выпилила exception handler

phareebo committed:
выпилила exception handler

phareebo committed:
выпилила exception handler

phareebo committed:
выпилила exception handler

phareebo committed:
рефакторинг

phareebo committed:
рефакторинг

phareebo committed:
рефакторинг

phareebo committed:
рефакторинг

phareebo committed:
рефакторинг

phareebo committed:
рефакторинг

Ramil23-bit committed:
added update workSchedule, getByIdWorkSchedule.

Commit:
BJS2-77981
Создание Skill

Ramil23-bit committed:
fixed code.

renamed variables in method checkCorrectedWorkSchedule.
Added the package to exception

phareebo committed:
@ToString.Exclude вынесен на уровень класса. В @EqualsAndHashCode исключены поля с двунаправленными связями

phareebo committed:
@ToString.Exclude вынесен на уровень класса. В @EqualsAndHashCode исключены поля с двунаправленными связями

phareebo committed:
@ToString.Exclude вынесен на уровень класса. В @EqualsAndHashCode исключены поля с двунаправленными связями

phareebo committed:
@ToString.Exclude вынесен на уровень класса. В @EqualsAndHashCode исключены поля с двунаправленными связями

phareebo committed:
@ToString.Exclude вынесен на уровень класса. В @EqualsAndHashCode исключены поля с двунаправленными связями

phareebo committed:
@ToString.Exclude вынесен на уровень класса. В @EqualsAndHashCode исключены поля с двунаправленными связями

phareebo committed:
вынесены: * документация из API
* валидация в отдельный класс

phareebo committed:
вынесены: * документация из API
* валидация в отдельный класс

phareebo committed:
вынесены: * документация из API
* валидация в отдельный класс

phareebo committed:
вынесены: * документация из API
* валидация в отдельный класс

phareebo committed:
маппинг в контроллере. маппинг листов в отдельном методе

phareebo committed:
маппинг в контроллере. маппинг листов в отдельном методе

phareebo committed:
маппинг в контроллере. маппинг листов в отдельном методе

phareebo committed:
маппинг в контроллере. маппинг листов в отдельном методе

phareebo committed:
маппинг в контроллере. маппинг листов в отдельном методе

phareebo committed:
маппинг в контроллере. маппинг листов в отдельном методе

phareebo committed:
метод получения события

phareebo committed:
метод получения события

phareebo committed:
метод получения события

phareebo committed:
метод получения события

phareebo committed:
метод получения события

phareebo committed:
метод получения события

phareebo committed:
метод поиска событий по фильтрам

phareebo committed:
метод поиска событий по фильтрам

phareebo committed:
метод поиска событий по фильтрам

phareebo committed:
метод поиска событий по фильтрам

phareebo committed:
метод поиска событий по фильтрам

phareebo committed:
метод поиска событий по фильтрам

phareebo committed:
условие поиска по фильтрам or -> and

phareebo committed:
метод удаления события

phareebo committed:
метод удаления события

phareebo committed:
метод удаления события

phareebo committed:
метод удаления события

phareebo committed:
метод удаления события

phareebo committed:
метод удаления события

phareebo committed:
метод получения событий пользователя

phareebo committed:
метод получения событий пользователя

phareebo committed:
метод получения событий пользователя

phareebo committed:
метод получения событий пользователя

phareebo committed:
метод получения событий пользователя

phareebo committed:
метод получения событий пользователя

phareebo committed:
метод обновления события; рефакторинг

phareebo committed:
метод обновления события; рефакторинг

phareebo committed:
метод обновления события; рефакторинг

phareebo committed:
метод обновления события; рефакторинг

phareebo committed:
метод обновления события; рефакторинг

phareebo committed:
метод обновления события; рефакторинг

Ramil23-bit committed:
fixed code.

fixed the update method, getting by id

Commit:
BJS2-77981
Получение списка скиллов пользователя

Keriod committed:
BJS2-77986. Подписка на других пользователей

1) PR правки

phareebo committed:
метод получения событий, в которых участвует пользователь

phareebo committed:
метод получения событий, в которых участвует пользователь

phareebo committed:
метод получения событий, в которых участвует пользователь

phareebo committed:
метод получения событий, в которых участвует пользователь

phareebo committed:
оставлен только 1 EventRequest

phareebo committed:
оставлен только 1 EventRequest

phareebo committed:
оставлен только 1 EventRequest

phareebo committed:
оставлен только 1 EventRequest

Keriod committed:
BJS2-77986. Подписка на других пользователей
- Изменение структуры пакетов
- Создание entity Subscription

Commit:
BJS2-77981
Получение списка предложенных скиллов пользователя.
Поправила получение скиллов: теперь вывод для текущего пользователя

phareebo committed:
userId из хедера, а не параметров

phareebo committed:
userId из хедера, а не параметров

phareebo committed:
userId из хедера, а не параметров

phareebo committed:
userId из хедера, а не параметров

phareebo committed:
рефакторинг директорий

phareebo committed:
рефакторинг директорий

phareebo committed:
рефакторинг директорий

phareebo committed:
рефакторинг директорий

dextomi committed:
Реализация основных операций по карьере

dextomi committed:
Реализация основных операций по карьере

Ramil23-bit commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,65 @@
+        career.setUser(user);
+        Career savedCareer = careerRepository.save(career);
+        return careerMapper.toCareerDto(savedCareer);
+    }
```
Дмитрий говорил что маппер в сервисе не используем. Он должен быть в контроллере

dextomi replied:
оке, перенес маппер в контроллер

Ramil23-bit commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,65 @@
+        validateDateFrom(careerDto);
+
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Пользователь не найден."));
```
Юзера можно искать через конфиг контекс, оттуда можно текущего пользователя вытащить, если тебе нужно

dextomi replied:
Не уверен что это нужно сейчас

Ramil23-bit commented on src/main/java/school/faang/user_service/mappers/career/CareerMapper.java:
```diff
@@ -0,0 +1,17 @@
+package school.faang.user_service.mappers.career;
+
```
с названием пакетов, лучше в единственном числе, чтобы потом не было конфликтов при Merge

dextomi replied:
Поправил название пакета на единственное число

Ramil23-bit commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,65 @@
+    private void validateDateFrom(CareerDto careerDto) {
+        if (!careerDto.getFrom().isBefore(LocalDate.now())) {
+            throw new DataValidationException("Дата начала карьеры не может быть больше текущей даты.");
+        }
```
валидацию выносили в отдельный пакет

dextomi replied:
Вынес

Ramil23-bit committed:
fixed code.

added new Dto,fixed mapper, service and controller

phareebo committed:
@data удалена с entity связанных с Event

phareebo committed:
@data удалена с entity связанных с Event

phareebo committed:
@data удалена с entity связанных с Event

phareebo committed:
@data удалена с entity связанных с Event

phareebo committed:
обновление ивента put -> patch

phareebo committed:
userId из хедера

phareebo committed:
@Transactional added

Commit:
BJS2-77981
UserSkillController создан для получения навыков пользователя
SkillMapper переписан без mapstruct

phareebo committed:
правки

phareebo committed:
правки

dextomi committed:
Перенес маппер из методов сервиса в контроллер;
Валидацию по дате вынес в отдельный пакет;
Изменил наименования пакета mappers на mapper.

dextomi committed:
Перенес маппер из методов сервиса в контроллер;
Валидацию по дате вынес в отдельный пакет;
Изменил наименования пакета mappers на mapper.

dextomi committed:
Перенес маппер из методов сервиса в контроллер;
Валидацию по дате вынес в отдельный пакет;
Изменил наименования пакета mappers на mapper.

20cognosce commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,32 @@
+import school.faang.user_service.mapper.career.CareerMapper;
+import school.faang.user_service.service.career.CareerService;
+
+@Controller
```
@RestController

20cognosce commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,32 @@
+
+@Controller
+@RequiredArgsConstructor
+public class CareerController {
```
@RequestMapping("/api/v1/user/career")

20cognosce replied:
Карьера это атрибут пользователя, я тебе не совсем точно написал базовый урл, наверно лучше 
/api/v1/user/{userId}/career

добавить и изменить можно только себе, а получить любого человека

20cognosce replied:
в общем все три урла детальнее обдумай

20cognosce commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,32 @@
+    private final CareerService careerService;
+    private final CareerMapper careerMapper;
+
+    public CareerDto addCareer(long userId, CareerDto careerDto) {
```
Эндпоинты создай, как ты через постман тестировал?

20cognosce replied:
На создание отдельная дто, на обновление отдельная и на ответ отдельная
3 ДТО

20cognosce commented on src/main/java/school/faang/user_service/dto/career/CareerDto.java:
```diff
@@ -0,0 +1,14 @@
+@Data
+public class CareerDto {
+    private long id;
+    private LocalDate from;
+    private LocalDate to;
+    private String company;
+    private String position;
```
на вход то что приходит должно валидироваться, используй аннотации

20cognosce replied:
ты не проверил, что даты адекватные, что компания и позиция не пустые

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,47 @@
+    private final UserRepository userRepository;
+    private final CareerRepository careerRepository;
+
+    public Career addCareer(long userId, Career career) {
```
@Transactional

20cognosce replied:
ты добавляешь карьеру самому себе, а не левому челу
поэтому userId не приходит с фронта, он приходит из заголовка запроса и доставать его из контекста надо как у Рамиля

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public Career updateCareer(long userId, Career career) {
+        CareerValidation.validateDateFrom(career);
```
Одной даты мало для проверки

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,47 @@
+            throw new DataValidationException("У вас нет прав изменять карьеру другого пользователя");
+        }
+
+        return careerRepository.save(existingCareer);
```
ну и что ты обновил из того, что в Career career тебе с фронта пришло? ничего
а надо проверить, какие поля не пустые, и их обновить

где тут обновления времени работы, компании, позиции - ничего нет

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,47 @@
+        return careerRepository.save(existingCareer);
+    }
+
+    public Career getById(long careerId) {
```
@Transactional(readOnly = true)

20cognosce reviewed (changes requested):
Ты должен Postman-ом протестировать свою задачу, и проверить что в базе создаются записи

Commit:
BJS2-77981
Приобретение предложенного скилла

Commit:
BJS2-77981
Использован существующий класс констант

Keriod committed:
BJS2-78003. Unit-тесты для задачи с подписками
- Тесты для service, filter, validator
- Правки по PR

Keriod committed:
BJS2-78003. Unit-тесты для задачи с подписками
- Тесты для service, filter, validator
- Правки по PR

Keriod committed:
BJS2-78003. Unit-тесты для задачи с подписками
- Тесты для service, filter, validator
- Правки по PR

Keriod committed:
BJS2-78003. Unit-тесты для задачи с подписками
- Тесты для service, filter, validator
- Правки по PR

Keriod committed:
BJS2-78003. Unit-тесты для задачи с подписками
- Тесты для service, filter, validator
- Правки по PR

Ramil23-bit committed:
participation in the event

implemented registration on event, cancellation of registration, obtaining a list of registered users and their number

dextomi committed:
Создал эндпоинты, добавил больше проверок для дат, для каждого метода создал отдельные ДТО, добавил в них же валидацию, сделал поиск юзера через userContext

dextomi committed:
Создал эндпоинты, добавил больше проверок для дат, для каждого метода создал отдельные ДТО, добавил в них же валидацию, сделал поиск юзера через userContext

Commit:
BJS2-77981
Optional добавлен в UserService в методе acquireSkillFromOffers.
В UserSkillController добавлен статус ответа PRECONDITION_FAILED

20cognosce commented:
Ну у тебя же конфликт на ПРе, почему не исправил, невозможно замерджить ветки

20cognosce commented on src/main/java/school/faang/user_service/controller/career/CareerController.java:
```diff
@@ -0,0 +1,56 @@
+    private final CareerService careerService;
+    private final CareerMapper careerMapper;
+
+    @PostMapping("/add")
```
в урле не должно быть действий, метод http итак отвечает на этот вопрос

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,69 @@
+            throw new SecurityException("Вы не можете редактировать записи о карьере другого пользователя.");
+        }
+
+        if (career.getDateFrom() != null) {
+            CareerValidation.validateDateFrom(existingCareer);
+            existingCareer.setDateFrom(career.getDateFrom());
+        }
+        if (career.getDateTo() != null) {
+            CareerValidation.validateDateTo(existingCareer);
+            existingCareer.setDateTo(career.getDateTo());
+        }
+        if (career.getCompany() != null) {
+            existingCareer.setCompany(career.getCompany());
+        }
+        if (career.getPosition() != null) {
+            existingCareer.setPosition(career.getPosition());
+        }
```
уже хорошо, но попробуй в приватный метод эту обобщенную проверку вынести, чтобы каждое поле сетилось в 1 строчку с проверкой на null и проверкой валидатора

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,69 @@
+        Career existingCareer = careerRepository.findById(career.getId())
+                .orElseThrow(() -> new EntityNotFoundException("Запись о карьере не найдена."));
+
+        if (!existingCareer.getUser().getId().equals(userId)) {
```
это не null-safe

20cognosce replied:
не исправлено

20cognosce commented on src/main/java/school/faang/user_service/validation/career/CareerValidation.java:
```diff
@@ -0,0 +1,34 @@
+        LocalDate currentDate = LocalDate.now();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd.MM.yyyy");
+
+        if (Stream.of(career.getDateFrom(), career.getDateTo())
+                .anyMatch(date -> date.isBefore(minDate) || date.isAfter(currentDate))) {
```
только в переменную вынеси, не надо стрим в if лепить

20cognosce commented on src/main/java/school/faang/user_service/validation/career/CareerValidation.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public static void validateDateRanges(Career career) {
+        LocalDate minDate = LocalDate.of(1970, 1, 1);
```
и для этого наверняка константа есть EPOCH или типо того

20cognosce reviewed (changes requested):
В целом уже норм

Keriod committed:
BJS2-78003. Unit-тесты для задачи с подписками
- PR правки

Keriod committed:
BJS2-78003. Unit-тесты для задачи с подписками
- PR правки

Keriod committed:
Removed git ignored files

Ramil23-bit committed:
fixed code

rename method in Service, fixed method in Controller

Commit:
BJS2-77981
GlobalExceptionHandler добавлен.
UserSkillController исправлен.
Язык исключений и логов - английский.
Exceptions в skillMapper.

Commit:
BJS2-77981
EntityNotFoundException удалён, вместо него используется RecordNotFoundException
SkillValidator перемещён

dextomi committed:
В валидации стрим в условии на проверку дат вынес в отдельную переменную;
В сервисе для метода updateCareer создал отдельный метод для обновления полей;
Поменял url в контроллере;
В CareerUpdateDto добавил аннотацию @NotNull для поля id.

dextomi committed:
Merge branch 'hippogriff-master-stream10' into hippogriff-fearture-stream10-BJS2-77994-career

20cognosce commented:
137 коммитов...

20cognosce commented on src/main/java/school/faang/user_service/controller/event/impl/EventControllerImpl.java:
```diff
@@ -0,0 +1,97 @@
+@RequiredArgsConstructor
+@Slf4j
+@RequestMapping("/api/v1/events")
+public class EventControllerImpl implements EventController {
```
Что это

20cognosce commented on src/main/java/school/faang/user_service/dto/event/request/EventRequest.java:
```diff
@@ -0,0 +1,68 @@
+@Builder
+@ValidEndDate
+@Validated
+public class EventRequest implements EndDateValidatable {
```
Что это

20cognosce commented on src/main/java/school/faang/user_service/service/career/CareerService.java:
```diff
@@ -0,0 +1,72 @@
+    private final CareerRepository careerRepository;
+    private final UserContext userContext;
+
+    private <T> void updateField(T newValue, Career existingCareer,
+                                 Consumer<Career> validator,
+                                 Consumer<T> setter) {
+        Optional.ofNullable(newValue).ifPresent(newFieldValue -> {
+            Optional.ofNullable(validator)
+                    .ifPresent(fieldValidator -> fieldValidator.accept(existingCareer));
+            setter.accept(newFieldValue);
+        });
+    }
```
Во-первых приватные методы внизу класса
Во-вторых чет слишком сложно написал, можно попроще и лишнее убрать? Зачем тебе newValue и existingCareer? Двух параметров достаточно - проверки и действия

20cognosce reviewed (changes requested):
Приведи ПР в порядок, можешь переоткрыть новый
